ReportTiming callback should record the sendFrameToEngine when it was scheduled

dkwingsmt opened the pull request:
Fixes https://github.com/flutter/flutter/issues/144261

## Pre-launch Checklist

- [ ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ ] I signed the [CLA].
- [ ] I listed at least one issue that this PR fixes in the description above.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [ ] I added new tests to check the change I am making, or this PR is [test-exempt].
- [ ] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [ ] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[test-exempt]: https://github.com/flutter/flutter/wiki/Tree-hygiene#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[Data Driven Fixes]: https://github.com/flutter/flutter/wiki/Data-driven-Fixes

dkwingsmt committed:
Add test

dkwingsmt committed:
Fix

dkwingsmt committed:
Better doc

goderbauer commented on packages/flutter/test/widgets/binding_live_test.dart:
```diff
@@ -0,0 +1,55 @@
+void main() {
+  LiveTestWidgetsFlutterBinding();
+  testWidgets('ReportTiming callback records the sendFramesToEngine when it was scheduled', (WidgetTester tester) async {
+    // This test needs LiveTestWidgetsFlutterBinding for multiple reasons.
+    // First, this was the environment that this bug was discovered.
+    // Second, LiveTestWidgetsFlutterBinding doesn't override
+    // scheduleWarmUpFrame, so that the test can start with no frames rendered
+    // allowing deferFirstFrame to work (see below).
```
I don't understand this paragraph. In my observation, the problem is that the warm-up frame triggered internally by `testWidgets` before the test body executes is actually send to the engine, but after that `resetFirstFrame` is called, which now puts engine and framework in an inconsistent state: The framework now things no frame has been sent (which means `deferFirstFrame` does have an effect) while the engine things the first frame has been sent.

dkwingsmt replied:
Kind of. The `_runTestBody` has:
```dart
    runApp(Container(key: UniqueKey(), child: _preTestMessage)); // Reset the tree to a known state.
    await pump();
```
so a frame is pumped to the engine anyway, warm-up or not. 

And your understanding on the inconsistency is absolutely correct. Now there's a reportTiming callback pending, which is supposed to only happen after rendered frames, while the framework thinks no frames have been rendered, allowing `deferFirstFrame` to take effect. This combined situation of `LiveTestWidgetsFlutterBinding` and `testWidgets` is probably the only case that can trigger this bug.

goderbauer commented on packages/flutter/test/widgets/binding_live_test.dart:
```diff
@@ -0,0 +1,55 @@
+    // the dummy widget is rendered, its reportTiming callback from the engine
+    // has not arrived yet.
+    //
+    // Note that the `runApp` call also schedules a warm up frame, which would
+    // have directly rendered the frame if the test had used
+    // `AutomatedTestWidgetsFlutterBinding`.
```
This doesn't seem super relevant and adds more confusion to what is actually going on here. Can we just drop this? Also note that "note that" is banned: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#avoid-empty-prose

goderbauer commented on packages/flutter/test/widgets/binding_live_test.dart:
```diff
@@ -0,0 +1,55 @@
+    // have directly rendered the frame if the test had used
+    // `AutomatedTestWidgetsFlutterBinding`.
+    expect(tester.binding.sendFramesToEngine, true);
+    // Push the widget with runApp instead of tester.pump, avoiding rendering a
+    // frame is rendered, which is needed for `deferFirstFrame` later to work.
```
This sentence is busted: "avoiding rendering a frame is rendered".

goderbauer commented on packages/flutter/test/widgets/binding_live_test.dart:
```diff
@@ -0,0 +1,55 @@
+    // Defer the first frame, making sendFramesToEngine false, and widget tree
+    // will be built but not sent to the engine.
+    tester.binding.deferFirstFrame();
+    // Pump a while, letting the reportTiming callback to run. If it had not
+    // recorded the `sendFramesToEngine` when it was scheduled, it would have
+    // crashed seeing `sendFramesToEngine` being now false!
+    await tester.pump(const Duration(milliseconds: 1));
```
Suggestion:

```suggestion
    expect(tester.binding.sendFramesToEngine, false);
    // Wait for the reportTiming callback (which completes the future below) to run. If the reportTiming callback were to assume that `sendFramesToEngine` is true, the callback would crash.
    await tester.binding.waitUntilFirstFrameRasterized;
```

goderbauer reviewed (commented):
Overall this seems like the right fix, the assert makes incorrect assumptions since we offer API to reset `sendFramesToEngine`.

dkwingsmt committed:
Resolve some comments

dkwingsmt committed:
Doc

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into record-send-frames

dkwingsmt commented:
@goderbauer I've addressed your feedbacks and rewrote some paragraphs. Can you take another look? (I'm filing an issue to record the problem, which I'll link to the test file. )

goderbauer commented on packages/flutter/test/widgets/binding_live_test.dart:
```diff
@@ -0,0 +1,61 @@
+    // `resetFirstFrameSent`, the framework thinks no frames have been rendered.
+
+    expect(tester.binding.sendFramesToEngine, true);
+    // Push the widget with runApp instead of tester.pump, avoiding rendering a
```
nitty-nit: For consistency with the other comments put runApp and tester.pump in ` as well?

goderbauer commented on packages/flutter/test/widgets/binding_live_test.dart:
```diff
@@ -0,0 +1,61 @@
+    runApp(const DummyWidget());
+    // Verify that no widget tree is built and nothing is rendered.
+    expect(find.text('First frame'), findsNothing);
+    // Defer the first frame, making sendFramesToEngine false, so that widget
```
same with sendFramesToEngine here

goderbauer reviewed (approved):
LGTM

goderbauer commented on packages/flutter/test/widgets/binding_live_test.dart:
```diff
@@ -0,0 +1,61 @@
+    // run. If the reportTiming callback were to assume that
+    // `sendFramesToEngine` is true, the callback would crash.
+    await tester.binding.waitUntilFirstFrameRasterized;
+    expect(find.text('First frame'), findsOne);
```
From the test failures: It looks like the test needs a `pump` call before this expect, which makes sense. Sorry to not include that in my previous suggestion.

dkwingsmt replied:
What do you think if I just revert to `    await tester.pump(const Duration(milliseconds: 1));`?

dkwingsmt replied:
I ended up including both.

dkwingsmt committed:
More comments

dkwingsmt commented:
I also wrote an issue to describe the problem https://github.com/flutter/flutter/issues/144261 and linked it to the test case. This PR will close the issue.

dkwingsmt committed:
Skip web

dkwingsmt committed:
Merge remote-tracking branch 'origin/main' into record-send-frames

dkwingsmt committed:
Fix compile

dkwingsmt committed:
Merge branch 'master' into record-send-frames
